[bug]: g++ 16 no longer defines __STDC_VERSION__ causing channelstorage_cpp_map_name_id.cc to fail

### Severity

Minor

### Versions

All

### Components/Modules

channelstorage_cpp_map_name_id.cc

### Operating Environment

Fedora 44, gcc 16.0.1

### Frequency of Occurrence

None

### Issue Description

include/asterisk/compat.h unconditionally checks __STDC_VERSION__ and sets it to 0 if it's not defined.  g++ 16.0 no longer sets that macro (which is actually the correct behavior) and attempting to set it manually causes a fatal error so channelstorage_cpp_map_name_id.cc fails to compile.


### Relevant log output

```shell
In file included from /usr/local/src/asterisk/asterisk/include/asterisk.h:22,
                 from channelstorage.h:26,
                 from channelstorage_cpp_map_name_id.cc:27:
/usr/local/src/asterisk/asterisk/include/asterisk/compat.h:27:9: error: ‘__STDC_VERSION__’ defined [-Werror]
   27 | #define __STDC_VERSION__ 0
      |         ^~~~~~~~~~~~~~~~
```

### Asterisk Issue Guidelines

- [x] Yes, I have read the Asterisk Issue Guidelines

--- a/include/asterisk/compat.h
+++ b/include/asterisk/compat.h
@@ -22,9 +22,11 @@
 
 #include "asterisk/compiler.h"
 
+#if !defined(__cplusplus) && !defined(c_plusplus)
 #ifndef __STDC_VERSION__
 /* flex output wants to find this defined. */
 #define	__STDC_VERSION__ 0
+#endif
 #endif
 
 #include <inttypes.h>

